opt: add rule to fold two grouping operators into one

DrewKimball opened the pull request:
Previously, the optimizer could not fold two grouping operators into
a single equivalent grouping operator.

This patch adds a rule that effects this transformation under the
following conditions:
1. All of the outer aggregates are aggregating on the output columns
   of the inner aggregates.
2. Every inner-outer aggregation pair can be replaced with an
   equivalent single aggregate.
3. The inner grouping columns functionally determine the outer
   grouping columns.
4. Both grouping operators are unordered.

As an example, the following query pairs are equivalent:
```
SELECT sum(t) FROM (SELECT sum(b) FROM ab GROUP BY a) AS g(t);
SELECT sum(b) FROM ab;

SELECT max(t) FROM (SELECT max(b) FROM ab GROUP BY a) AS g(t);
SELECT max(b) FROM ab;

SELECT sum_int(t) FROM (SELECT count(b) FROM ab GROUP BY a) AS g(t);
SELECT count(b) FROM ab;
```
This situation is rare in direct SQL queries, but can arise when
composing views and queries.

Release note (sql change): The optimizer can now fold two grouping
operators together when they are aggregating over functions like sum.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49627)
<!-- Reviewable:end -->

mgartner reviewed (approved):
Very cool! I left a few questions for my own understanding. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

It's outside the scope of this PR, but I think there's another case that can be folded: when there is an outer aggregate function (like max or min) on the grouped inner column. For example:

```
SELECT max(a), sum(b) FROM (SELECT x, sum(y) FROM r GROUP BY x) AS s(a, b)
=>
SELECT max(x), sum(y) FROM r
```

Not sure if this case is worth handling in the future, but thought I'd point it out.

> Reviewed 1 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-:-M8QrqAf8JJP5hailgmb:bnu9wzg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[pkg/sql/opt/operator.go, line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8Qm7WYDzl4MHQhzOtC:-M8Qm7WZ4v8wsK_ciRs6:b23w65) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1fec1fae4ec484079ef40605e0ba84dff3ce5d/pkg/sql/opt/operator.go#L374)):*
> ```Go
> 
> 	case CountOp, CountRowsOp:
> 		return outer == SumIntOp
> ```

So you're only checking for `SumIntOp` here because `outer` could never be a `SumOp` because `CountOp` and `CountRowsOp` always result in integers. Is my understanding correct?
___
*[pkg/sql/opt/norm/groupby_funcs.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8QqrCJ60xz7UC_4VGH:-M8QqrCKDBofWU9iEmx-:bwi40ij) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1fec1fae4ec484079ef40605e0ba84dff3ce5d/pkg/sql/opt/norm/groupby_funcs.go#L321)):*
> ```Go
> 		innerAgg := innerAggs[i].Agg
> 		if !opt.IsAggregateOp(innerAgg) {
> 			return false
> ```

When would an innerAgg or outerAgg be something other than an AggregateOp?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/operator.go, line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8Qm7WYDzl4MHQhzOtC:-M8RG2h36Fl3lppEdTav:btih660) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1fec1fae4ec484079ef40605e0ba84dff3ce5d/pkg/sql/opt/operator.go#L374)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

So you're only checking for `SumIntOp` here because `outer` could never be a `SumOp` because `CountOp` and `CountRowsOp` always result in integers. Is my understanding correct?
</blockquote></details>

Yes, exactly.
___
*[pkg/sql/opt/norm/groupby_funcs.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8QqrCJ60xz7UC_4VGH:-M8RGCGf65wZWOsmWoLH:barr143) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1fec1fae4ec484079ef40605e0ba84dff3ce5d/pkg/sql/opt/norm/groupby_funcs.go#L321)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

When would an innerAgg or outerAgg be something other than an AggregateOp?
</blockquote></details>

They could be AggFilters or AggDistincts. These could be allowed in the future but I wanted to keep things simple for this PR.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
Nice work. I wonder what the motivation for this was, did you encounter a case like this?

Also, I want to commend you on your use of "effects" haha

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-:-M8ROJqr5LbCRgLxJXKH:b8hcgv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @mgartner)

___
*[pkg/sql/opt/norm/groupby_funcs.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8QqrCJ60xz7UC_4VGH:-M8RLY7Z1qMyglp1A93W:bgisa1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1fec1fae4ec484079ef40605e0ba84dff3ce5d/pkg/sql/opt/norm/groupby_funcs.go#L321)):*
<details><summary><i>Previously, DrewKimball (Andrew Kimball) wrote…</i></summary><blockquote>

They could be AggFilters or AggDistincts. These could be allowed in the future but I wanted to keep things simple for this PR.
</blockquote></details>

It's worth adding a comment, mentioning AggFilters/AggDistinct (or checking for those particular ops) to make things more clear.
___
*[pkg/sql/opt/norm/rules/groupby.opt, line 378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8RNl_38Pv45MtUnWcL:-M8RNl_48_kV_uF1Llq3:b-jzm1hd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1fec1fae4ec484079ef40605e0ba84dff3ce5d/pkg/sql/opt/norm/rules/groupby.opt#L378)):*
> ```
> # ignore orderings).
> [FoldGroupingOperators, Normalize]
> (GroupBy | ScalarGroupBy | DistinctOn
> ```

We document FirstAgg as only being used in DistinctOn. I don't think we can execbuild an expression that has a FirstAgg inside a GroupBy. Conversely, DistinctOn doesn't support anything other than ConstAgg and FirstAgg. I think we should remove FirstAgg and DistinctOn altogether from this rule.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/norm/groupby_funcs.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8QqrCJ60xz7UC_4VGH:-M8Rb3lr3EFXsijdJfHY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1fec1fae4ec484079ef40605e0ba84dff3ce5d/pkg/sql/opt/norm/groupby_funcs.go#L321)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's worth adding a comment, mentioning AggFilters/AggDistinct (or checking for those particular ops) to make things more clear.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/groupby.opt, line 378 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8RNl_38Pv45MtUnWcL:-M8RbHkS26lljxo4NxcN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c1fec1fae4ec484079ef40605e0ba84dff3ce5d/pkg/sql/opt/norm/rules/groupby.opt#L378)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We document FirstAgg as only being used in DistinctOn. I don't think we can execbuild an expression that has a FirstAgg inside a GroupBy. Conversely, DistinctOn doesn't support anything other than ConstAgg and FirstAgg. I think we should remove FirstAgg and DistinctOn altogether from this rule.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
> Nice work. I wonder what the motivation for this was, did you encounter a case like this?
Once the joins are removed from the query at line 593 in xform/testdata/external/trading, we end up with a ScalarGroupBy on top of a GroupBy.

mgartner reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
> It's outside the scope of this PR, but I think there's another case that can be folded: when there is an outer aggregate function (like max or min) on the grouped inner column. For example:
> 
> ```
> SELECT max(a), sum(b) FROM (SELECT x, sum(y) FROM r GROUP BY x) AS s(a, b)
> =>
> SELECT max(x), sum(y) FROM r
> ```
> 
> Not sure if this case is worth handling in the future, but thought I'd point it out.

That's a good idea, I hadn't thought of that. Now that you mention it, I think you could also fold any outer aggregate on the grouped inner column if it has an AggDistinct.

mgartner commented:
> That's a good idea, I hadn't thought of that. Now that you mention it, I think you could also fold any outer aggregate on the grouped inner column if it has an AggDistinct.

I think that makes sense. Do you mean something like this:

```
SELECT count(DISTINCT a), sum(b) FROM (SELECT x, sum(y) FROM xy GROUP BY x) AS ab(a, b)
=>
SELECT count(DISTINCT x), sum(y) FROM xy
```

DrewKimball commented:
> I think that makes sense. Do you mean something like this:
> 
> ```
> SELECT count(DISTINCT a), sum(b) FROM (SELECT x, sum(y) FROM xy GROUP BY x) AS ab(a, b)
> =>
> SELECT count(DISTINCT x), sum(y) FROM xy
> ```

Yes, exactly. Although it's possible that there's some rule that would remove the AggDistinct.

rytaft reviewed (commented):
Nice! Just a few nits/questions.

> Reviewed 3 of 7 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-:-M8W7wd_9I5pcF-hBtZH:b-hgdoas)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)

___
*[pkg/sql/opt/operator.go, line 358 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8VzJon9cO8Ist1h5jI:-M8VzJon9cO8Ist1h5jJ:b-ctk94q) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/operator.go#L358)):*
> ```Go
> //
> //   SELECT sum(s) FROM (SELECT sum(y) FROM xy GROUP BY x) AS f(s);
> //   SELECT sum(y) FROM xyz;
> ```

[nit] xyz -> xy (here and below)
___
*[pkg/sql/opt/operator.go, line 374 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W304k-_rTc2axB7p3:-M8W304k-_rTc2axB7p4:b-84xdex) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/operator.go#L374)):*
> ```Go
> 
> 	case CountOp, CountRowsOp:
> 		return outer == SumIntOp
> ```

Is there a reason it can't also be `SumOp`?
___
*[pkg/sql/opt/norm/groupby_funcs.go, line 357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W4Fq6720iZbClwQCu:-M8W4Fq6720iZbClwQCv:b1a9rra) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/norm/groupby_funcs.go#L357)):*
> ```Go
> 
> // MergeAggs returns an AggregationsExpr that is equivalent to the two given
> // AggregationsExprs.
> ```

[nit] add a note that this will panic if `CanMergeAggs` is false.
___
*[pkg/sql/opt/norm/groupby_funcs.go, line 370 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W4fOs0tWpsVXW9ZZD:-M8W4fOs0tWpsVXW9ZZE:bxyiwqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/norm/groupby_funcs.go#L370)):*
> ```Go
> 		// For each outer aggregate, construct a new aggregate that takes the Agg
> 		// field of the referenced inner aggregate and the Col field of the outer
> 		// aggregate.
> ```

[nit] can you explain why you need the Agg from the inner and Col from the input to the outer?
___
*[pkg/sql/opt/norm/rules/groupby.opt, line 357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W5qFRAxFh9AlT0jwI:-M8W5qFRAxFh9AlT0jwJ:b-qsbmga) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/norm/rules/groupby.opt#L357)):*
> ```
> #   6 6 5
> #
> # Its functional dependencies: key(r), r->(a, b, c), a->(b, c)
> ```

c is not included in the table -- these FDs seem wrong
___
*[pkg/sql/opt/norm/testdata/rules/groupby, line 3023 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W7UV7A8203wJ_ZbAG:-M8W7UV7A8203wJ_ZbAH:b-ak3ruq) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/norm/testdata/rules/groupby#L3023)):*
> ```
> # Case with a GroupBy on a GroupBy.
> norm expect=FoldGroupingOperators
> SELECT sum_int(s) FROM (SELECT y, sum_int(x) AS s FROM xy GROUP BY y) GROUP BY y
> ```

Do you have any tests where the inner and outer group by columns are different but the inner still determines the outer? How about a test with more than one group by column?


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 1 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/operator.go, line 358 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8VzJon9cO8Ist1h5jI:-M8WXjmd--nSKo6YrQmq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/operator.go#L358)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] xyz -> xy (here and below)
</blockquote></details>

Done.
___
*[pkg/sql/opt/operator.go, line 374 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W304k-_rTc2axB7p3:-M8WXl7eAaMCGgVhe0kb:b-ni4sv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/operator.go#L374)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is there a reason it can't also be `SumOp`?
</blockquote></details>

CountOp and CountRowsOp always produce ints, while (I believe) SumOp produces a decimal number. So, replacing a SumOp on a CountOp would change the outputted type from decimal to int. I'll add a comment.
___
*[pkg/sql/opt/norm/groupby_funcs.go, line 357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W4Fq6720iZbClwQCu:-M8WYOoa0i9_JQ4o1wJq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/norm/groupby_funcs.go#L357)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] add a note that this will panic if `CanMergeAggs` is false.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/groupby_funcs.go, line 370 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W4fOs0tWpsVXW9ZZD:-M8WZLiABizlATVkRQ1W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/norm/groupby_funcs.go#L370)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] can you explain why you need the Agg from the inner and Col from the input to the outer?
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/groupby.opt, line 357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W5qFRAxFh9AlT0jwI:-M8WZZMz7-sqAy6hmIJr:bwulvkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/norm/rules/groupby.opt#L357)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

c is not included in the table -- these FDs seem wrong
</blockquote></details>

Right, I forgot to prune those after removing the c column from the table. Done.
___
*[pkg/sql/opt/norm/testdata/rules/groupby, line 3023 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W7UV7A8203wJ_ZbAG:-M8W_fGO2uOVCB2sD5LD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/norm/testdata/rules/groupby#L3023)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do you have any tests where the inner and outer group by columns are different but the inner still determines the outer? How about a test with more than one group by column?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/operator.go, line 374 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8W304k-_rTc2axB7p3:-M8XI-2cBmxjokOphcBq:bgh3h2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/09ea764acaa55a0aa8d00fddd41b58c0a5ab4d16/pkg/sql/opt/operator.go#L374)):*
<details><summary><i>Previously, DrewKimball (Andrew Kimball) wrote…</i></summary><blockquote>

CountOp and CountRowsOp always produce ints, while (I believe) SumOp produces a decimal number. So, replacing a SumOp on a CountOp would change the outputted type from decimal to int. I'll add a comment.
</blockquote></details>

Doh. I left the same question yesterday... I should have suggested a comment then. This is definitely comment-worthy, so thanks for adding it!


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 6 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-:-M8jzKAmBVZch1uEjNKC:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball)

___
*[pkg/sql/opt/norm/groupby_funcs.go, line 373 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8jyFVLEicDNv8hkGyo:-M8jyFVLEicDNv8hkGyp:bhpok9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5b20d3538515a147a258043c6a24734b3450183/pkg/sql/opt/norm/groupby_funcs.go#L373)):*
> ```Go
> 		// every inner-outer aggregate pair forms a valid decomposition for the
> 		// inner aggregate. The column from the outer aggregate has to be used to
> 		// preserve logical equivalency.
> ```

It would also help to add a reminder that you use the inner Agg field since both inner and outer are the same in most cases, but for count/count rows you need the inner. You can also add a reference to see `opt.AggregatesCanMerge` for details.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt: add rule to fold two grouping operators into one

Previously, the optimizer could not fold two grouping operators into
a single equivalent grouping operator.

This patch adds a rule that effects this transformation under the
following conditions:
1. All of the outer aggregates are aggregating on the output columns
   of the inner aggregates.
2. Every inner-outer aggregation pair can be replaced with an
   equivalent single aggregate.
3. The inner grouping columns functionally determine the outer
   grouping columns.
4. Both grouping operators are unordered.

As an example, the following query pairs are equivalent:
```
SELECT sum(t) FROM (SELECT sum(b) FROM ab GROUP BY a) AS g(t);
SELECT sum(b) FROM ab;

SELECT max(t) FROM (SELECT max(b) FROM ab GROUP BY a) AS g(t);
SELECT max(b) FROM ab;

SELECT sum_int(t) FROM (SELECT count(b) FROM ab GROUP BY a) AS g(t);
SELECT count(b) FROM ab;
```
This situation is rare in direct SQL queries, but can arise when
composing views and queries.

Release note (sql change): The optimizer can now fold two grouping
operators together when they are aggregating over functions like sum.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=49627) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=49627) before we can accept your contribution.<br/><hr/>**Drew Kimball** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=49627) it.</sub>

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/norm/groupby_funcs.go, line 373 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8jyFVLEicDNv8hkGyo:-M8kZxPkAH5E82hd86Vk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5b20d3538515a147a258043c6a24734b3450183/pkg/sql/opt/norm/groupby_funcs.go#L373)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

It would also help to add a reminder that you use the inner Agg field since both inner and outer are the same in most cases, but for count/count rows you need the inner. You can also add a reference to see `opt.AggregatesCanMerge` for details.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball)

___
*[pkg/sql/opt/norm/groupby_funcs.go, line 374 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8kgZE-ED1H-Xaj-ZJG:-M8kgZE03VZSl8OKs1zC:b-prhoot) ([raw file](https://github.com/cockroachdb/cockroach/blob/3806f861cad05ca8d72ff86b362fa5d2dec8475b/pkg/sql/opt/norm/groupby_funcs.go#L374)):*
> ```Go
> 		// inner aggregate. In most cases, the inner and outer aggregates are the
> 		// same, but in the count and count-rows cases the inner aggregate must
> 		// be used (see opt.aggregatesCanMerge for details). The column from the
> ```

[nit] opt.aggregatesCanMerge -> opt.AggregatesCanMerge


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/norm/groupby_funcs.go, line 374 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49627#-M8kgZE-ED1H-Xaj-ZJG:-M8kkjZn9j5WNLLEA5mZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3806f861cad05ca8d72ff86b362fa5d2dec8475b/pkg/sql/opt/norm/groupby_funcs.go#L374)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] opt.aggregatesCanMerge -> opt.AggregatesCanMerge
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49627)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
bors r+
